rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,206 @@
+---
+authors: Andrew Lytvynov (andrew@gravitational.com)
+state: draft
+---
+
+# RFD 15 - 2FA device management
+
+## What
+
+Improvements to 2FA device management (OTP and U2F):
+- support for multiple 2FA devices per account
+- ability to mix OTP and U2F devices (and possibly more in the future)
+- ability to list/add/delete devices by the user from Web UI and CLI
+
+## Why
+
+Currently, Teleport only supports a single OTP or U2F device per local user.
+This setting is global to the cluster - either everyone uses OTP or U2F or
+nothing. Users don't have a choice, and losing a 2FA device requires an admin
+to reset the account to recover.
+
+[RFD 14](0014-session-2FA.md) adds support for 2FA challenges per-connection.
+2FA devices will become more widely used by Enterprise users, who would
+previously delegate this to their SSO provider.
+
+Teleport needs to make 2FA device support more flexible, as both a
+long-standing OSS users' request, and the new Enterprise use-case.
+
+## Details
+
+### protocols
+
+Teleport supports two 2FA protocols: OTP and U2F.
+
+There's no change to the list of supported protocols, but the implementation
+should assume that more protocols will be added (e.g. WebAuthn).
+
+For U2F, we will migrate from https://github.com/tstranex/u2f to
+https://github.com/flynn/u2f, to allow client-side CLI authentication without
+the `u2f-host` dependency.
+
+### UX
+
+For the below examples, we assume that the user already has at least one 2FA
+device enrolled. The [bootstrap](#bootstrap) section covers registration of the
+first device.
+
+#### CLI
+
+Login:
+
+```sh
+$ tsh login --login=non-sso-user
+Enter password for Teleport user non-sso-user: ...
+Tap your security key... <tap>
+
+$ tsh login --login=non-sso-user --2fa=otp
+Enter password for Teleport user non-sso-user: ...
+Enter OTP code: ...
+
+$ tsh login --login=sso-user --auth=oidc
+# SSO page opens
+# no 2FA prompt from Teleport
+```
+
+2FA management:
+
+```sh
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh 2fa add
+Adding a new 2FA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+2FA device "solokey" added.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
```

Comment:
shall we remove by ID instead of the name to avoid name collisions and other possible problems?

Reply:
IDs are not very user-friendly.
We will prevent users from creating devices with existing names.
All these names are namespaced under the user anyway (see the storage section below), so there's no conflicts between users.

Reply:
We can follow github strategy where they can type first 5 letters and copy them. I just think that having user-input data as a key is dangerous for so many reasons :)

Reply:
You'd probably can still delete by ID, but let users delete by name.

Reply:
It wouldn't be a key in backend storage path, since keys get stored in the user object.
I don't want to add IDs on top of this unless necessary, to keep things simple.

Reply:
how would you trace the key usage in the certificate? could it be public key fingerprint used as an ID? I still think that using user-input data will be a problem

Reply:
I don't think we need to specify which specific key was used in the certificate.
Let me sketch out the cert content changes.

Reply:
This could be helpful for audit purposes, because we will be able to connect action to a specific device. We will gain ability to block the activity by device that got lost without blocking the user.

Reply:
I get a YAGNI smell from these key IDs, but adding regardless. It's only a small amount of additional complexity.

Added cert extensions and key ID in https://github.com/gravitational/teleport/pull/5125/commits/d26feeb64c69e43c486bd7c5783b411e2e4f95c6